### ISSUE TITLE
Collapse/un freezes when there are a lot of children nodes

### ISSUE THREAD

**Issue by AboutC0DE** [Lodin/react-vtree#30]
I used https://github.com/Lodin/react-vtree/blob/HEAD/__stories__/VariableSizeTree.story.tsx example and just changed data generator: 

```
const createNode = (depth: number = 0) => {
  const node: DataNode = {
    children: [],
    id: nodeId,
    name: `test-${nodeId}`,
  }

  nodeId += 1

  if (depth === **2**) {
    return node
  }

  for (let i = 0; i < **1000**; i++) {
    node.children.push(createNode(depth + 1))
  }

  return node
}
```

**Comment by AlokBisht** [comment]
I'm facing the same issue. when there are too many nodes in the tree, and all are expanded. We traverse the child nodes if a parent is open but we don't consider if the nodes will be visible or hidden cause of the overflow. even though the the tree may show a small number of nodes at a time, traversing the whole hierarchy is not efficient. I think the yield should happen only when requested by view (or make it use choice).

**Comment by Lodin** [comment]
@AlokBisht, yeah, I already had that discussion in the #11. The main problem with `yield` that is cannot work backward (or at least I cannot imagine how to set it up). So if you are scrolling down, it's ok. But when you start scrolling up the whole system breaks.

I probably should re-open #11 to make it easier to find. Let's continue the discussion in that issue.

**Comment by Lodin** [comment]
Ok, I think I've solved this issue. For the @akarabach's example results are quite nice. Unfortunately, cannot check something bigger because Chrome fails with out of memory error. 

@AlokBisht, @akarabach, you can check [fix/performance-on-big-data](https://github.com/Lodin/react-vtree/tree/fix/performance-on-big-data) branch if you have time. I updated storybook examples, so it should be easy to follow all the changes (they are mostly internal though the `treeWalker` function has changed a lot).

**Comment by Lodin** [comment]
@akarabach, @AlokBisht, all changes are published now in [3.0.0-beta.0](https://github.com/Lodin/react-vtree/releases/tag/3.0.0-beta.0) release.
